forecast.ts

```diff
@@ -8,227 +9,386 @@ const jiraPassword = process.env.JIRA_PASSWORD;
+  : undefined;
 
 const jira = new JiraApi({
   protocol: jiraProtocol,
   host: jiraHost,
   port: jiraPort,
   username: jiraUsername,
   password: jiraPassword,
-  apiVersion: '2',
-  strictSSL: true
+  apiVersion: "2",
+  strictSSL: true,
 });
 
 type TicketResponse = {
-    readonly total: number;
-    readonly issues: ReadonlyArray<string>;
+  readonly total: number;
+  readonly issues: ReadonlyArray<string>;
 };
 
-const issuesForSearchQuery = async (searchQuery: string, maxResults: number = 1000): Promise<TicketResponse> => {
-    const issuesResp = await jira.searchJira(searchQuery, { maxResults });
-    return parseJiraResponse(issuesResp);
+// convert provided time interval into weeks
+const durationInWeeks =
+  timeUnit === "weeks" ? timeLength : timeLength / daysInWeek;
+
+/**
+ * Collects issues from JIRA to analyse and facilitate prediction.
+ *
+ * @param searchQuery Query to retrieve data from JIRA.
+ * @param maxResults Maximum number of results to retrieve.
+ * @returns The tickets retrieved from JIRA.
+ */
+const issuesForSearchQuery = async (
+  searchQuery: string,
+  maxResults: number = 1000
+): Promise<TicketResponse> => {
+  const issuesResp = await jira.searchJira(searchQuery, { maxResults });
+  return parseJiraResponse(issuesResp);
 };
 
 // TODO: It would be better to use the date QA was completed for the ticket instead of the date the
 //       ticket was resolved.
-const fetchResolvedTicketsPerSprint = async () => {
-    // We want to know how many tickets were completed during each sprint. To make things easier,
-    // we're defining a sprint as just any period of two weeks.
-    let historyStart = -sprintLengthInWeeks;
-    let historyEnd = 0;
-    const ticketCounts: Promise<number>[] = [];
-
-    while (historyStart >= -1 * numWeeksOfHistory) {
-        const query =
-          `project = ${jiraProjectID} AND issuetype in standardIssueTypes() AND issuetype != Epic ` +
-          `AND resolved >= ${historyStart}w AND resolved <= ${historyEnd}w`;
-
-        ticketCounts.push(
-            issuesForSearchQuery(query, 0).then(r => r.total)
-        );
-
-        historyStart -= sprintLengthInWeeks;
-        historyEnd -= sprintLengthInWeeks;
-    }
+/**
+ * Gets tickets resolved in each time interval cycle.
+ *
+ * @returns An array of number of tickets resolved in each time interval.
+ */
+const fetchResolvedTicketsPerTimeInterval = async () => {
+  // We want to know how many tickets were completed during each time interval. If not defined,
+  // our time interval is just any period of two weeks.
+  let historyStart = -durationInWeeks;
+  let historyEnd = 0;
+  const ticketCounts: Promise<number>[] = [];
 
-    return Promise.all(ticketCounts);
-};
+  while (historyStart >= -1 * numWeeksOfHistory) {
+    const query =
+      `project = ${jiraProjectID} AND issuetype in standardIssueTypes() AND issuetype != Epic ` +
+      `AND resolved >= ${historyStart}w AND resolved <= ${historyEnd}w`;
```

Comment:
Would it be worth using days (presumably `d`) instead of weeks (`w`) in this query so `TIME_LENGTH` doesn't have to be a multiple of 7 if they're using `TIME_UNIT=days`? Maybe as a separate ticket/PR if it's not trivial?

Reply:
Yes, we can just swap out `durationInWeeks` with `durationInDays` and do the conversion the other way! Shouldn't be too hard, trying it on this PR.